File: app/src/main/java/com/ggslk/ggslk/adapter/CategoryRecyclerAdapter.java
package com.ggslk.ggslk.adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;
import com.ggslk.ggslk.R;
import com.ggslk.ggslk.model.Category;
import com.squareup.picasso.Picasso;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;

public class CategoryRecyclerAdapter extends RecyclerView.Adapter<CategoryRecyclerAdapter.CategoryViewHolder> {
    private List<Category> categories;
    private RequestQueue mRequestQueue;

    public CategoryRecyclerAdapter(List<Category> categories) {
        this.categories = categories;
    }

    @Override
    public CategoryViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.recyclerview_category, parent, false);
        CategoryViewHolder categoryViewHolder = new CategoryViewHolder(v);

        // RequestQueue initialized
        mRequestQueue = Volley.newRequestQueue(v.getContext());

        return categoryViewHolder;
    }

    @Override
    public void onBindViewHolder(CategoryViewHolder holder, int position) {
        holder.categoryArticleName.setText(categories.get(position).getFeaturedArticle().getTitle());
        holder.categoryArticleAuthor.setText(categories.get(position).getFeaturedArticle().getAuthor().getName());
        holder.categoryTitle.setText(categories.get(position).getTitle());
    }

    @Override
    public int getItemCount() {
        return categories.size();
    }

    @Override
    public void onBindViewHolder(@NonNull final CategoryViewHolder holder, final int position, @NonNull List<Object> payloads) {
        super.onBindViewHolder(holder, position, payloads);

        mRequestQueue.add(new JsonObjectRequest
                (Request.Method.GET, "https://ggslk.com/api/get_category_posts?slug=" + categories.get(position).getSlug() + "&count=1", null, new Response.Listener<JSONObject>() {
                    @Override
                    public void onResponse(JSONObject response) {
                        try {
                            String title = response.getJSONArray("posts").getJSONObject(0)
                                    .get("title").toString();
                            String author = response.getJSONArray("posts").getJSONObject(0)
                                    .getJSONObject("author").get("name").toString();
                            String imageUrl = response.getJSONArray("posts").getJSONObject(0)
                                    .getJSONObject("thumbnail_images").getJSONObject("full").get("url").toString();

                            categories.get(position).getFeaturedArticle().setTitle(title);
                            categories.get(position).getFeaturedArticle().getAuthor().setName(author);
                            categories.get(position).getFeaturedArticle().setImageUrl(imageUrl);

                            Picasso.get().load(imageUrl).fit().centerCrop().into(holder.categoryArticleImageView);
                            holder.categoryArticleName.setText(title);
                            holder.categoryArticleAuthor.setText(author);

                        } catch (JSONException e) {
                            e.printStackTrace();
                        }

                    }
                }, new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        error.printStackTrace();
                    }
                }));
    }

    class CategoryViewHolder extends RecyclerView.ViewHolder {
        private CardView cardView;
        private ImageView categoryArticleImageView;
        private TextView categoryArticleName;
        private TextView categoryArticleAuthor;
        private TextView categoryTitle;

        public CategoryViewHolder(View itemView) {
            super(itemView);
            cardView = itemView.findViewById(R.id.categoryCardView);
            categoryArticleImageView = itemView.findViewById(R.id.categoryArticleImageView);
            categoryArticleName = itemView.findViewById(R.id.categoryArticleNameTextView);
            categoryArticleAuthor = itemView.findViewById(R.id.categoryArticleAuthorTextView);
            categoryTitle = itemView.findViewById(R.id.categoryTitleTextView);
        }
    }
}

File: app/src/main/java/com/ggslk/ggslk/adapter/CategoryRecyclerAdapter.kt
package com.ggslk.ggslk.adapter

import android.content.Context
import android.support.constraint.ConstraintLayout
import android.support.v4.app.FragmentManager
import android.support.v7.widget.CardView
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.android.volley.DefaultRetryPolicy
import com.android.volley.Request
import com.android.volley.Response
import com.android.volley.toolbox.JsonObjectRequest
import com.ggslk.ggslk.R
import com.ggslk.ggslk.common.Session
import com.ggslk.ggslk.fragment.CategoryArticlesFragment
import com.ggslk.ggslk.model.Article
import com.ggslk.ggslk.model.Author
import com.ggslk.ggslk.model.Category
import com.squareup.picasso.Picasso
import org.json.JSONException

class CategoryRecyclerAdapter(private val context: Context, private val categories: List<Category>, private val fragmentManager: FragmentManager) : RecyclerView.Adapter<CategoryRecyclerAdapter.CategoryViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoryViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.recyclerview_category, parent, false)
        return CategoryViewHolder(v)
    }

    override fun onBindViewHolder(holder: CategoryViewHolder, position: Int) {
        holder.cardViewBackground.setBackgroundColor(Session.categoryColors[position % Session.categoryColors.size])
        holder.category = categories[position]
        holder.categoryTitle.text = categories[position].title

        var featuredArticle = categories[position].featuredArticle
        if (featuredArticle != null) {
            Picasso.get().load(featuredArticle.imageUrl).fit().centerCrop().into(holder.categoryArticleImageView)
            holder.categoryArticleName.text = featuredArticle.title
            holder.categoryArticleAuthor.text = featuredArticle.author!!.name
        } else {
            val jsonObjectRequest = JsonObjectRequest(Request.Method.GET, "https://ggslk.com/api/get_category_posts?slug=" + categories[position].slug + "&count=1", null, Response.Listener { response ->
                try {
                    val firstPost = response.getJSONArray("posts").getJSONObject(0)

                    var author = Author()
                    author.name = firstPost.getJSONObject("author").get("name").toString()

                    var article = Article()
                    article.title = firstPost.get("title").toString()
                    article.imageUrl = firstPost.getJSONObject("thumbnail_images").getJSONObject("full").get("url").toString()
                    article.author = author

                    categories[position].featuredArticle = article

                    Picasso.get().load(article.imageUrl).fit().centerCrop().into(holder.categoryArticleImageView)
                    holder.categoryArticleName.text = article.title
                    holder.categoryArticleAuthor.text = author.name
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
            }, Response.ErrorListener { error -> error.printStackTrace() })

            jsonObjectRequest.retryPolicy = DefaultRetryPolicy(500000,
                    DefaultRetryPolicy.DEFAULT_MAX_RETRIES,
                    DefaultRetryPolicy.DEFAULT_BACKOFF_MULT)

            jsonObjectRequest.tag = "cat"

            Session.mRequestQueue!!.add(jsonObjectRequest)
        }
    }

    override fun getItemCount(): Int {
        return categories.size
    }

    inner class CategoryViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var category: Category? = null

        val cardView: CardView = itemView.findViewById(R.id.categoryCardView)
        val cardViewBackground: ConstraintLayout = itemView.findViewById(R.id.categoryCardViewBackgroundLayout)
        val categoryArticleImageView: ImageView = itemView.findViewById(R.id.categoryArticleImageView)
        val categoryArticleName: TextView = itemView.findViewById(R.id.categoryArticleNameTextView)
        val categoryArticleAuthor: TextView = itemView.findViewById(R.id.categoryArticleAuthorTextView)
        val categoryTitle: TextView = itemView.findViewById(R.id.categoryTitleTextView)

        init {
            cardView.setOnClickListener {
                val transaction = fragmentManager.beginTransaction()
                val fragment = CategoryArticlesFragment.newInstance(category!!.slug!!)
                transaction.replace(R.id.container, fragment)
                transaction.commit()
            }
        }
    }
}
